File: backend/app/services/neuropsicologos_service.py
import logging
from typing import Optional
from .database_service import DatabaseService

logger = logging.getLogger(__name__)


class NeuropsicologosService:
    """Servicio para listar neuropsicólogos usando psycopg2.

    Nota: En ausencia de una tabla específica de neuropsicólogos, se obtiene
    desde la tabla `usuario` filtrando por rol `Neuropsicólogo`.
    Si tu esquema incluye una tabla dedicada (p. ej. `neuropsicologo`),
    podemos ajustar los SELECT y JOIN según corresponda.
    """

    def __init__(self):
        self.db = DatabaseService()

    def get_role_id_by_name(self, role_name: str) -> Optional[int]:
        try:
            query = "SELECT id_rol FROM rol WHERE nom_rol = %s LIMIT 1"
            with self.db.get_cursor() as cursor:
                cursor.execute(query, (role_name,))
                row = cursor.fetchone()
                return row["id_rol"] if row else None
        except Exception as e:
            logger.error(f"Error obteniendo id_rol por nombre {role_name}: {e}")
            return None

    def get_all(self, page: int = 1, limit: int = 5, search: Optional[str] = None, include_inactive: bool = True, role_id: Optional[int] = None):
        try:
            offset = (page - 1) * limit

            # Determinar role_id a usar (Neuropsicólogo por defecto)
            effective_role_id = role_id
            if effective_role_id is None:
                # Intentar obtener por nombre; fallback a 3 si no existe
                rid = self.get_role_id_by_name('Neuropsicólogo')
                effective_role_id = rid if rid is not None else 3

            # Condiciones
            where = ["u.id_rol = %s"]
            params = [effective_role_id]

            if not include_inactive:
                where.append("r.estado_rol = true")

            if search:
                where.append("(u.usua ILIKE %s)")
                params.append(f"%{search}%")

            where_clause = f"WHERE {' AND '.join(where)}" if where else ""

            # Consulta principal: mapeo mínimo para frontend
            query = f"""
                SELECT 
                    -- Campos esperados por el frontend
                    u.id_usuario AS id_neuropsicologo,
                    u.usua::text AS nombres,
                    ''::text AS apellidos,
                    u.usua AS username,
                    ''::text AS email,
                    COALESCE(r.estado_rol, true) AS estado,
                    NOW()::timestamp AS fecha_registro,

                    -- Campos reales de la tabla usuario y rol (para gestión)
                    u.id_usuario AS id_usuario,
                    u.id_rol AS id_rol,
                    u.usua AS usua,
                    r.nom_rol AS rol_nombre,
                    r.estado_rol AS estado_rol
                FROM usuario u
                JOIN rol r ON u.id_rol = r.id_rol
                {where_clause}
                ORDER BY u.usua ASC
                LIMIT %s OFFSET %s
            """

            main_params = params + [limit, offset]

            count_query = f"""
                SELECT COUNT(*) AS count
                FROM usuario u
                JOIN rol r ON u.id_rol = r.id_rol
                {where_clause}
            """

            with self.db.get_cursor() as cursor:
                cursor.execute(query, main_params)
                rows = cursor.fetchall() or []
                data = [dict(r) for r in rows]

                cursor.execute(count_query, params)
                total = cursor.fetchone()["count"] if cursor.rowcount != 0 else 0

                total_pages = (total + limit - 1) // limit

                return {
                    "success": True,
                    "data": data,
                    "metadata": {
                        "total": total,
                        "page": page,
                        "limit": limit,
                        "total_pages": total_pages,
                        "has_next": page < total_pages,
                        "has_prev": page > 1,
                    },
                }
        except Exception as e:
            logger.error(f"Error obteniendo neuropsicólogos: {e}")
            return {
                "success": False,
                "message": f"Error al obtener neuropsicólogos: {str(e)}",
                "data": [],
                "metadata": {
                    "total": 0,
                    "page": page,
                    "limit": limit,
                    "total_pages": 0,
                    "has_next": False,
                    "has_prev": False,
                },
            }
    
    def get_total_neuropsicologos(self):
        """Obtener el total de neuropsicólogos activos"""
        try:
            # Obtener el ID del rol de Neuropsicólogo
            role_id = self.get_role_id_by_name('Neuropsicólogo')
            if not role_id:
                role_id = 3  # Fallback
            
            query = """
                SELECT COUNT(*) as total 
                FROM usuario u 
                JOIN rol r ON u.id_rol = r.id_rol 
                WHERE u.id_rol = %s AND r.estado_rol = true
            """
            with self.db.get_cursor() as cursor:
                cursor.execute(query, (role_id,))
                result = cursor.fetchone()
                return result[0] if result else 0
        except Exception as e:
            logger.error(f"Error obteniendo total de neuropsicólogos: {e}")
            return 0

File: frontend/src/services/neuropsicologosService.ts
import { useEffect, useState } from 'react'
import { apiClient } from './api'

const RESOURCE = '/auth/obtener_neuropsicologo'

export interface Neuropsicologo {
  id_usuario: number;
  id_rol: number;
  usua: string;
  contra: string;
  estado_usuario: number;
}

export interface NeuropsicologoCreate {
  nombres: string
  apellidos: string
  username: string
  estado?: boolean
}

export interface NeuropsicologoUpdate {
  nombres?: string
  apellidos?: string
  username?: string
  estado?: boolean
}

export interface NeuropsicologosResponse {
  success: boolean
  data: Neuropsicologo[]
  metadata: {
    total: number
    page: number
    limit: number
    total_pages: number
    has_next: boolean
    has_prev: boolean
  }
  message?: string
}

export interface NeuropsicologoResponse {
  success: boolean
  data?: Neuropsicologo
  message: string
}

export function useGetNeuropsicologos() {
  const [neuropsicologos, setNeuropsicologos] = useState<Neuropsicologo[]>([])
  const [metadata, setMetadata] = useState({
    total: 0,
    page: 1,
    limit: 5,
    total_pages: 0,
    has_next: false,
    has_prev: false,
  })
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState<string | null>(null)



  // Fallback-aware fetch: intenta /neuropsicologos y si no existe usa /users?role_id=3
  const fetchNeuropsicologosSafe = async (
    page: number = 1,
    limit: number = 5,
    search: string = '',
    includeInactive: boolean = true,
  ) => {
    try {
      setLoading(true)
      setError(null)
      try {
        const { data } = await apiClient.get<NeuropsicologosResponse>(RESOURCE, {
          params: { page, limit, ...(search ? { search } : {}), ...(includeInactive ? { include_inactive: true } : {}) },
        })
        if (data?.success) {
          setNeuropsicologos(data.data || [])
          setMetadata(data.metadata)
          return
        }
      } catch (err) {
        // Fallback a endpoint placeholder disponible en backend
        const { data } = await apiClient.get<any>('/users-db', { params: { page, limit, ...(search ? { search } : {}) } })
        const mapped: Neuropsicologo[] = (data?.data || []).map((u: any) => ({
          id_neuropsicologo: u.id ?? u.id_usuario ?? 0,
          nombres: u.nombres ?? u.name ?? u.usua ?? '',
          apellidos: u.apellidos ?? '',
          username: u.username ?? u.usua ?? u.name ?? '',
          email: u.email ?? '',
          estado: u.estado ?? true,
          fecha_registro: u.createdAt ?? u.fecha_registro,
        }))
        setNeuropsicologos(mapped)
        setMetadata({
          total: data?.total ?? mapped.length,
          page: data?.page ?? page,
          limit: data?.limit ?? limit,
          total_pages: data?.totalPages ?? Math.ceil((data?.total ?? mapped.length) / (data?.limit ?? limit)),
          has_next: (data?.page ?? page) < (data?.totalPages ?? 1),
          has_prev: (data?.page ?? page) > 1,
        })
      }
    } catch (e) {
      setError('Error al cargar neuropsicólogos')
    } finally {
      setLoading(false)
    }
  }

  useEffect(() => {
    fetchNeuropsicologosSafe(1, 5)
  }, [])

  return { neuropsicologos, metadata, loading, error, refetch: fetchNeuropsicologosSafe, setNeuropsicologos }
}

export const neuropsicologosService = {
  async getAll(page = 1, limit = 10, search = '', includeInactive = true): Promise<NeuropsicologosResponse> {
    const { data } = await apiClient.get<NeuropsicologosResponse>(RESOURCE, {
      params: { page, limit, ...(search ? { search } : {}), ...(includeInactive ? { include_inactive: true } : {}) },
    })
    return data
  },
  async getById(id: number): Promise<NeuropsicologoResponse> {
    const { data } = await apiClient.get<NeuropsicologoResponse>(`${RESOURCE}/${id}`)
    return data
  },
  async create(payload: NeuropsicologoCreate): Promise<NeuropsicologoResponse> {
    const { data } = await apiClient.post<NeuropsicologoResponse>(RESOURCE, payload)
    return data
  },
  async update(id: number, payload: NeuropsicologoUpdate): Promise<NeuropsicologoResponse> {
    const { data } = await apiClient.put<NeuropsicologoResponse>(`${RESOURCE}/${id}`, payload)
    return data
  },
  async delete(id: number): Promise<NeuropsicologoResponse> {
    const { data } = await apiClient.delete<NeuropsicologoResponse>(`/auth/eliminar_neuropsicologo/${id}`)
    return data
  },
  async restore(id: number): Promise<NeuropsicologoResponse> {
    const { data } = await apiClient.post<NeuropsicologoResponse>(`${RESOURCE}/${id}/restore`)
    return data
  },
}


